### PR TITLE
added scanf family

### PR DESCRIPTION
**Pull request by maheitsec**
Added 6 functions:
scanf, fscanf, sscanf, vscanf, vfscanf, vsscanf
Please review and merge.

### CODE REVIEW

**Comment by riataman** [comment]
__func__ needs to go.  The name of the function is different in macOS.

**Comment by ronaldtse** [comment]
@maheitsec thanks for this. We'd rather not have this stale, perhaps you could update per the reviews so we can merge this soon?

**Comment by maheitsec** [comment]
@drystone I fixed scanf family functions for C90. Now I'm making ./test/scanf.c for testing. But I don't know well about cmockatest. Could you please explain? How can I see the retrace log?

**Comment by maheitsec** [comment]
@riataman Thanks for your checking. I replaced func as __func__. Is it Ok in MacOS?

**Comment by drystone** [comment]
@maheitsec there are some examples in the comckatest.c Tracing goes to stderr so you can replace stderr. Something like

```
FILE *oldsderr, *f;
int fd[2];
pipe(fd);
oldstderr = stderr;
stderr = fdopen(fd[1]), "w");
do_stuff();
fclose(f);
stderr = oldstderr;
char buf[1024];
read(fd[0], buf, 1024);
close(fd[0]);
strncmp(test_str, buf, strlen(test_str));
```
Most times I've just checked the return value from the rtr_{func} is the same as that from {func} but the more tests the better!

File: scanf.c [hunk 1]
```diff
@@ -146,12 +164,16 @@ RETRACE_REPLACE(vsscanf)
...
+	rtr_fileno_t fileno_;
 	va_list ap1;
+	int result;
+
+	real_vfscanf = RETRACE_GET_REAL(vfscanf);
+	fileno_ = RETRACE_GET_REAL(fileno);
 
-	va_copy(ap1, ap);
-	int result = real_vfscanf(stream, format, ap);
+	__va_copy(ap1, ap);
```
Comment (drystone) [inline scanf.c]: What's the reason for `__va_copy()`?

**Comment by riataman** [comment]
@maheitsec: Also please add whatever code you are using to test as a standalone c file. check in tests for example id.c, env.c and add it to runtests.sh.

None of the cmocka tests work in macOS and they are very limited in what they test (most of the stuff in there just tests that the funcion exists in the .so, which is of limited use).

Would be nice to have a proper testing setup. I think cmocka is making this harder here, not easier. I would be nice to have something like ctests where we can run the binary with retrace loaded and compare output vs a regexp.

File: scanf.c (hunk 1 above)
Comment (maheitsec) [inline scanf.c]: va_copy() occurs compile error in current Makefile (I think C90), so I replaced it to \_\_va_copy().

**Comment by drystone** [comment]
@riataman why don't the cmocka tests work in MacOS? I remember you saying `fmfmopen()` isn't supported but that can easily be replaced by `pipe()` And the existing tests are just skeletons waiting for some flesh :)

**Comment by riataman** [comment]
@drystone: Well I guess the cmoka stuff could be making to work with some effort. BUT I don't think those tests are actually testing much of our real functionality.

First and foremost the cmoka stuff is "dlopening" retrace so the all the functions aren't really begin replaced, this fact just basically leaves 99% of stuff I have been running into untested (loops, deadlocks etc).

I would rather just replace the cmoka stuff with "ctest" (https://cmake.org/Wiki/CMake/Testing_With_CTest) and use PASS_REGULAR_EXPRESSION tests, where I run a binary under the real conditions retrace is going to be used so I that I can actually catch all loops, crashes, etc.

**Comment by drystone** [comment]
@riataman the cmocka tests are unit tests and should probably stay but I
see your point regarding real world tests.

**Comment by maheitsec** [comment]
@drystone Hello, I added testing modules. Could you please check?

**Comment by drystone** [comment]
@maheitsec I thought it looked good but there are a load of formatting
problems being flagged - also the linux Travis job seems to have
hung so It'll need some investigation (did you see my comment about automated
tests - it's no good if the test program is waiting for keyoard input!). I'm
out for a few hours but will look this evening. Maybe you'll have fixed it in
the meantime :)

```

WARNING: space prohibited between function name and open parenthesis '('RNING: space prohibited between function name and open parenthesis '('
#45: FILE: test/cmockatest.c:718:
+fp1 = fopen ("scanf_test1.txt", "w+");

ERROR: spaces required around that '=' (ctx:WxV)
#48: FILE: test/cmockatest.c:721:
+r1 =fscanf(fp1, "%s", str1);
    ^

WARNING: space prohibited between function name and open parenthesis '('
#50: FILE: test/cmockatest.c:723:
+fp2 = fopen ("scanf_test2.txt", "w+");

ERROR: "foo * bar" should be "foo *bar"
#72: FILE: test/cmockatest.c:745:
+to_vscanf(rtr_vscanf_t fn, const char * fmt, ...)

WARNING: Missing a blank line after declarations
#75: FILE: test/cmockatest.c:748:
+	va_list ap;
+	va_start(ap, fmt);

WARNING: Missing a blank line after declarations
#77: FILE: test/cmockatest.c:750:
+	int result = fn(fmt, ap);
+	va_end(ap);

WARNING: Missing a blank line after declarations
#98: FILE: test/cmockatest.c:771:
+	va_list ap;
+	va_start(ap, fmt);

WARNING: Missing a blank line after declarations
#100: FILE: test/cmockatest.c:773:
+	int result = fn(stream, fmt, ap);
+	va_end(ap);

WARNING: space prohibited between function name and open parenthesis '('
#109: FILE: test/cmockatest.c:782:
+fp1 = fopen ("scanf_test1.txt", "w+");

WARNING: space prohibited between function name and open parenthesis '('
#114: FILE: test/cmockatest.c:787:
+fp2 = fopen ("scanf_test2.txt", "w+");

WARNING: Missing a blank line after declarations
#127: FILE: test/cmockatest.c:800:
+	va_list ap;
+	va_start(ap, fmt);

WARNING: Missing a blank line after declarations
#129: FILE: test/cmockatest.c:802:
+	int result = fn(str, fmt, ap);
+	va_end(ap);

WARNING: please, no spaces at the start of a line
#154: FILE: test/cmockatest.c:871:
+      cmocka_unit_test(test_rtr_scanf), cmocka_unit_test(test_rtr_fscanf),$

WARNING: please, no spaces at the start of a line
#155: FILE: test/cmockatest.c:872:
+      cmocka_unit_test(test_rtr_sscanf), cmocka_unit_test(test_rtr_vscanf),$

WARNING: please, no spaces at the start of a line
#156: FILE: test/cmockatest.c:873:
+      cmocka_unit_test(test_rtr_vfscanf), cmocka_unit_test(test_rtr_vsscanf),$

ERROR: Bad function definition - int scanf_test() should probably be int scanf_test(void)
#171: FILE: test/scanf.c:30:
+int scanf_test()

ERROR: Bad function definition - int fscanf_test() should probably be int fscanf_test(void)
#188: FILE: test/scanf.c:44:
+int fscanf_test()

ERROR: "foo * bar" should be "foo *bar"
#192: FILE: test/scanf.c:48:
+	FILE * fp;

WARNING: space prohibited between function name and open parenthesis '('
#194: FILE: test/scanf.c:50:
+	fp = fopen ("scanf_test.txt", "w+");

ERROR: space prohibited before that close parenthesis ')'
#200: FILE: test/scanf.c:56:
+	printf("Read String1 |%s|\n", str1 );

ERROR: space prohibited before that close parenthesis ')'
#201: FILE: test/scanf.c:57:
+	printf("Read String2 |%s|\n", str2 );

ERROR: space prohibited before that close parenthesis ')'
#202: FILE: test/scanf.c:58:
+	printf("Read String3 |%s|\n", str3 );

ERROR: space prohibited before that close parenthesis ')'
#203: FILE: test/scanf.c:59:
+	printf("Read Integer |%d|\n", year );

ERROR: Bad function definition - int sscanf_test() should probably be int sscanf_test(void)
#210: FILE: test/scanf.c:66:
+int sscanf_test()

ERROR: space prohibited after that open parenthesis '('
#215: FILE: test/scanf.c:71:
+	strcpy( dtm, "Saturday March 25 1989" );

ERROR: space prohibited before that close parenthesis ')'
#215: FILE: test/scanf.c:71:
+	strcpy( dtm, "Saturday March 25 1989" );

ERROR: space prohibited after that open parenthesis '('
#216: FILE: test/scanf.c:72:
+	sscanf( dtm, "%s %s %d  %d", weekday, month, &day, &year );

ERROR: space prohibited before that close parenthesis ')'
#216: FILE: test/scanf.c:72:
+	sscanf( dtm, "%s %s %d  %d", weekday, month, &day, &year );

WARNING: unchecked sscanf return value
#216: FILE: test/scanf.c:72:
+	sscanf( dtm, "%s %s %d  %d", weekday, month, &day, &year );

ERROR: space prohibited before that close parenthesis ')'
#218: FILE: test/scanf.c:74:
+	printf("%s %d, %d = %s\n", month, day, year, weekday );

ERROR: "foo * bar" should be "foo *bar"
#223: FILE: test/scanf.c:79:
+void GetMatchesVscanf ( const char * format, ... )

ERROR: space prohibited after that open parenthesis '('
#223: FILE: test/scanf.c:79:
+void GetMatchesVscanf ( const char * format, ... )

ERROR: space prohibited before that close parenthesis ')'
#223: FILE: test/scanf.c:79:
+void GetMatchesVscanf ( const char * format, ... )

WARNING: Missing a blank line after declarations
#226: FILE: test/scanf.c:82:
+	va_list args;
+	va_start (args, format);

WARNING: space prohibited between function name and open parenthesis '('
#226: FILE: test/scanf.c:82:
+	va_start (args, format);

WARNING: space prohibited between function name and open parenthesis '('
#228: FILE: test/scanf.c:84:
+	va_end (args);

ERROR: Bad function definition - int vscanf_test() should probably be int vscanf_test(void)
#231: FILE: test/scanf.c:87:
+int vscanf_test ()

WARNING: space prohibited between function name and open parenthesis '('
#231: FILE: test/scanf.c:87:
+int vscanf_test ()

WARNING: space prohibited between function name and open parenthesis '('
#236: FILE: test/scanf.c:92:
+	printf ("Please enter a number and a word: ");

WARNING: space prohibited between function name and open parenthesis '('
#237: FILE: test/scanf.c:93:
+	fflush (stdout);

WARNING: space prohibited between function name and open parenthesis '('
#239: FILE: test/scanf.c:95:
+	printf ("Number read: %d\nWord read: %s\n", val, str);

ERROR: "foo * bar" should be "foo *bar"
#244: FILE: test/scanf.c:100:
+void ReadStuff (FILE * stream, const char * format, ...)

ERROR: "foo * bar" should be "foo *bar"
#244: FILE: test/scanf.c:100:
+void ReadStuff (FILE * stream, const char * format, ...)

WARNING: space prohibited between function name and open parenthesis '('
#244: FILE: test/scanf.c:100:
+void ReadStuff (FILE * stream, const char * format, ...)

WARNING: Missing a blank line after declarations
#247: FILE: test/scanf.c:103:
+	va_list args;
+	va_start (args, format);

WARNING: space prohibited between function name and open parenthesis '('
#247: FILE: test/scanf.c:103:
+	va_start (args, format);

WARNING: space prohibited between function name and open parenthesis '('
#249: FILE: test/scanf.c:105:
+	va_end (args);

ERROR: Bad function definition - int vfscanf_test() should probably be int vfscanf_test(void)
#252: FILE: test/scanf.c:108:
+int vfscanf_test ()

WARNING: space prohibited between function name and open parenthesis '('
#252: FILE: test/scanf.c:108:
+int vfscanf_test ()

ERROR: "foo * bar" should be "foo *bar"
#254: FILE: test/scanf.c:110:
+	FILE * pFile;

WARNING: space prohibited between function name and open parenthesis '('
#258: FILE: test/scanf.c:114:
+	pFile = fopen ("myfile.txt","r");

ERROR: space required after that ',' (ctx:VxV)
#258: FILE: test/scanf.c:114:
+	pFile = fopen ("myfile.txt","r");
 	                           ^

ERROR: spaces required around that '!=' (ctx:VxV)
#260: FILE: test/scanf.c:116:
+	if (pFile!=NULL) {
 	         ^

WARNING: space prohibited between function name and open parenthesis '('
#261: FILE: test/scanf.c:117:
+		ReadStuff ( pFile, " %s %d ", str, &val );

ERROR: space prohibited after that open parenthesis '('
#261: FILE: test/scanf.c:117:
+		ReadStuff ( pFile, " %s %d ", str, &val );

ERROR: space prohibited before that close parenthesis ')'
#261: FILE: test/scanf.c:117:
+		ReadStuff ( pFile, " %s %d ", str, &val );

WARNING: space prohibited between function name and open parenthesis '('
#262: FILE: test/scanf.c:118:
+		printf ("I have read %s and %d", str, val);

WARNING: space prohibited between function name and open parenthesis '('
#263: FILE: test/scanf.c:119:
+		fclose (pFile);

ERROR: "foo * bar" should be "foo *bar"
#269: FILE: test/scanf.c:125:
+void GetMatchesSscanf ( const char * str, const char * format, ... )

ERROR: "foo * bar" should be "foo *bar"
#269: FILE: test/scanf.c:125:
+void GetMatchesSscanf ( const char * str, const char * format, ... )

ERROR: space prohibited after that open parenthesis '('
#269: FILE: test/scanf.c:125:
+void GetMatchesSscanf ( const char * str, const char * format, ... )

ERROR: space prohibited before that close parenthesis ')'
#269: FILE: test/scanf.c:125:
+void GetMatchesSscanf ( const char * str, const char * format, ... )

WARNING: Missing a blank line after declarations
#272: FILE: test/scanf.c:128:
+	va_list args;
+	va_start (args, format);

WARNING: space prohibited between function name and open parenthesis '('
#272: FILE: test/scanf.c:128:
+	va_start (args, format);

WARNING: space prohibited between function name and open parenthesis '('
#274: FILE: test/scanf.c:130:
+	va_end (args);

ERROR: Bad function definition - int vsscanf_test() should probably be int vsscanf_test(void)
#277: FILE: test/scanf.c:133:
+int vsscanf_test ()

WARNING: space prohibited between function name and open parenthesis '('
#277: FILE: test/scanf.c:133:
+int vsscanf_test ()

ERROR: space prohibited after that open parenthesis '('
#282: FILE: test/scanf.c:138:
+	GetMatchesSscanf ( "99 bottles of beer on the wall", " %d %s ", &val, buf);

WARNING: space prohibited between function name and open parenthesis '('
#284: FILE: test/scanf.c:140:
+	printf ("Product: %s\nQuantity: %d\n", buf, val);

ERROR: Bad function definition - int main() should probably be int main(void)
#289: FILE: test/scanf.c:145:
+int main()

total: 34 errors, 36 warnings, 281 lines checked

**Comment by maheitsec** [comment]
@drystone Could you please do auto-check again? And do you have an idea for testing scanf function without input?

**Comment by drystone** [comment]
@maheitsec in a similar way to the example I gave you yesterday. scanf() reads from stdin.

```
FILE *oldstdin;
char buf[1024];
int fd[2];
pipe(fd);
oldstdin = stdin;
stdin = fdopen(fd[0]), "r");
write(fd[1], "string you want to test");
scanf("%s", buf);
fclose(stdin);
stdin = oldstdin;
close(fd[1]);
```

**Comment by maheitsec** [comment]
@drystone Hello, I fixed scanf test functions. Could you please check?

**Comment by maheitsec** [comment]
@drystone Hello I have an error in Travis ci.
ERROR: trailing whitespace
#26: FILE: test/scanf.c:100:
+^I$
How can I fix this?

**Pull request closed without merging**